### PR TITLE
Enable rule.unused-paramater

### PR DESCRIPTION
**Pull request by Gusted**
- As title, it enables a new linter rule.
- Over the few weeks that I've traversed trough the code-base I've seen a little bit _too many_ issues that were caused by not used the correct paramaters, this could be prevented if a rule was erroring that the code has a unused paramater. With this rule you can be sure that you only use the paramaters that are needed, and within code-review, the reviewer knows that a certain paramater is not used within the scope of that function.
- Most are replaced with `_` some of them are "fixed". Added few `//FIXME/TODO` for things that I saw while going to a bit too many files.

Just to note, I'm still using `_` and not completely removing the parameter, as if I was reading such code it makes it more clear that it won't be used.

### CODE REVIEW

**Comment by wxiaoguang** [comment]
For code like `setup("hooks/post-receive.log", c.Bool("debug"))`, I think we should keep them and improve them, instead of removing the parameter.

**Comment by Gusted** [comment]
> For code like `setup("hooks/post-receive.log", c.Bool("debug"))`, I think we should keep them and improve them, instead of removing the parameter.

Yeah, its a bit of a harder task to do so. As you will now include file logging in the console logging as a optional thing. I wasn't able to do within 15 minutes, which is why its now a removed parameter.

**Comment by wxiaoguang** [comment]
And I worry about we shouldn't just remove these unused parameters. 

Some unused parameters may indicate some old bugs.

File: modules/git/diff.go [hunk 1]
```diff
@@ -293,7 +293,7 @@ func GetAffectedFiles(oldCommitID, newCommitID string, env []string, repo *Repos
 	err = NewCommand("diff", "--name-only", oldCommitID, newCommitID).
 		RunInDirTimeoutEnvFullPipelineFunc(env, -1, repo.Path,
 			stdoutWriter, nil, nil,
-			func(ctx context.Context, cancel context.CancelFunc) error {
+			func(_ context.Context, _ context.CancelFunc) error {
```
Comment (wxiaoguang) [inline modules/git/diff.go:296]: For example, I am not sure whether it is correct to ignore  the `cancel context.CancelFunc`, usually it might be called somewhere.

So maybe we could spend enough time to review the suspicious usages one by one.

**Comment by Gusted** [comment]
> And I worry about we shouldn't just remove these unused parameters.
 Some unused parameters may indicate some old bugs.

Damn, you're quick at spotting the real reason of this PR - I totally agree with you!

So, I first wanted to open up a issue about it that we could investigate the unused paramaters and see if they should be used. Yet nobody wants to scroll trough a list of 500 errors. So this PR was created so anyone can take a quick look and see if they recognize anything.

I don't even expect this PR to be merged as a whole, but rather as a "mega-thread" for possible bugs and mini-PR's whose fixes bugs that are now discovered.

File: modules/markup/common/footnote.go [hunk 2]
```diff
@@ -232,7 +232,7 @@ func (b *footnoteBlockParser) Open(parent ast.Node, reader text.Reader, pc parse
 	return item, parser.HasChildren
 }
 
-func (b *footnoteBlockParser) Continue(node ast.Node, reader text.Reader, pc parser.Context) parser.State {
+func (b *footnoteBlockParser) Continue(_ ast.Node, reader text.Reader, _ parser.Context) parser.State {
```
Comment (wxiaoguang) [inline modules/markup/common/footnote.go:235]: Here `Continue(node ast.Node` and above `Open(parent ast.Node`, after the refactoring, it seems more difficult to understand.

Is it possible to use `_node ast.Node` and `_parent ast.Node` to pass the lint?

File: modules/notification/action/action.go [hunk 3]
```diff
@@ -402,10 +403,10 @@ func (a *actionNotifier) NotifySyncCreateRef(doer *user_model.User, repo *models
 	}
 }
 
-func (a *actionNotifier) NotifySyncDeleteRef(doer *user_model.User, repo *models.Repository, refType, refFullName string) {
+func (a *actionNotifier) NotifySyncDeleteRef(doer *user_model.User, repo *models.Repository, _, refFullName string) {
 	if err := models.NotifyWatchers(&models.Action{
-		ActUserID: repo.OwnerID,
-		ActUser:   repo.MustOwner(),
+		ActUserID: doer.ID,
+		ActUser:   doer,
```
Comment (wxiaoguang) [inline modules/notification/action/action.go:409]: Is it a bug fix? If yes, we may need a separate PR.

File: modules/markup/common/footnote.go (hunk 2 above)
Comment (Gusted) [reply modules/markup/common/footnote.go:235]: > Is it possible to use _node ast.Node and _parent ast.Node to pass the lint?

```
modules/markup/common/footnote.go:286:32: parameter '_node' seems to be unused, consider removing or renaming it as _
```

Unfortunately not, but we can make use of here by removing `_` such that only the type is displayed. Or we could use `//revive:disable.....`
But that is just more hacks IMO

Comment (wxiaoguang) [reply modules/markup/common/footnote.go:235]: Well .... maybe we can suggest the lint authors to accept such styles .... let lint users decide whether they accept `_var` variables.

File: modules/notification/action/action.go (hunk 3 above)
Comment (Gusted) [reply modules/notification/action/action.go:409]: Not really - the passed argument `doer` is already the owner(in a lot of cases doer can be removed as it's present on things like issue and comments). So it's just "refactor" and use the already present parameter which already is the owner.

Comment (wxiaoguang) [reply modules/notification/action/action.go:409]: I haven't read related code, but I just remember many bugs are caused by incorrect `owner`/`doer` usage (I also have used them wrongly before). In many cases, `doer` is not `owner`, for example, an admin user can do everything, then the doer should be the admin user, not the owner.

File: modules/markup/common/footnote.go (hunk 2 above)
Comment (silverwind) [reply modules/markup/common/footnote.go:235]: Agree, would prefer `_node` over `_` because it preserves the meaning while marking the parameter unused. We have this exact setup also for JS:

https://github.com/go-gitea/gitea/blob/49933c7fb09cc98ae0e57a483d68011ac67fd492/.eslintrc#L305

Is it possible to configure the go linter like that as well?

Comment (Gusted) [reply modules/markup/common/footnote.go:235]: Yeah that would makes sense, that it could be configured within the settings, I've opened a Issue for it https://github.com/mgechev/revive/issues/613

File: modules/notification/action/action.go (hunk 3 above)
Comment (Gusted) [reply modules/notification/action/action.go:409]: > an admin user can do everything, then the doer should be the admin user, not the owner.

In this case, the only usage of the code is with `xxx.MustOwner()`(which I assume is the correct doer) passed into `doer`, but it would be a bug fix if the usage would pop somewhere else up whereby it's by a repo admin.

File: modules/notification/webhook/webhook.go [hunk 4]
```diff
@@ -769,12 +769,12 @@ func sendReleaseHook(doer *user_model.User, rel *models.Release, action api.Hook
 		return
 	}
 
-	mode, _ := models.AccessLevel(rel.Publisher, rel.Repo)
+	mode, _ := models.AccessLevel(doer, rel.Repo)
```
Comment (Gusted) [inline modules/notification/webhook/webhook.go:772]: @wxiaoguang This is a bug - the rel.Publisher doesn't get changed between "edits" and "delete" of the release:

```go
func (m *webhookNotifier) NotifyNewRelease(rel *models.Release) {
	sendReleaseHook(rel.Publisher, rel, api.HookReleasePublished)
}

func (m *webhookNotifier) NotifyUpdateRelease(doer *user_model.User, rel *models.Release) {
	sendReleaseHook(doer, rel, api.HookReleaseUpdated)
}

func (m *webhookNotifier) NotifyDeleteRelease(doer *user_model.User, rel *models.Release) {
	sendReleaseHook(doer, rel, api.HookReleaseDeleted)
}
```

Guess this will deserve it's own PR?

File: modules/session/db.go [hunk 5]
```diff
@@ -91,7 +91,7 @@ type DBProvider struct {
 
 // Init initializes DB session provider.
 // connStr: username:password@protocol(address)/dbname?param=value
-func (p *DBProvider) Init(maxLifetime int64, connStr string) error {
+func (p *DBProvider) Init(maxLifetime int64, _ string) error {
```
Comment (Gusted) [inline modules/session/db.go:94]: This also looks suspicious given the comment above it.

File: modules/notification/webhook/webhook.go (hunk 4 above)
Comment (Gusted) [reply modules/notification/webhook/webhook.go:772]: #17806

File: modules/session/db.go (hunk 5 above)
Comment (lunny) [reply modules/session/db.go:94]: Yeah, the parameter meaning lost.

File: modules/markup/common/footnote.go (hunk 2 above)
Comment (Gusted) [reply modules/markup/common/footnote.go:235]: > Well .... maybe we can suggest the lint authors to accept such styles .... let lint users decide whether they accept `_var` variables.

Not much luck on that one.

```suggestion
func (b *footnoteBlockParser) Continue(ast.Node, reader text.Reader, parser.Context) parser.State {
```
I think such thing can be used to have the readability.
```

File: modules/session/db.go (hunk 5 above)
Comment (chavacava) [reply modules/session/db.go:94]: It is the comment that is misleading because `connStr` has no effect in the method behavior (i.e. the parameter means nothing to this method)

File: modules/markup/common/footnote.go (hunk 2 above)
Comment (chavacava) [reply modules/markup/common/footnote.go:235]:
```go
func (b *footnoteBlockParser) Continue(ast.Node, reader text.Reader, parser.Context) parser.State 
```

I'm not sure the compiler lets you mix named and unnamed parameters

File: modules/session/db.go (hunk 5 above)
Comment (zeripath) [reply modules/session/db.go:94]: Yeah the comment is old.

File: models/branches.go [hunk 6]
```diff
@@ -581,7 +580,7 @@ func RemoveDeletedBranch(repoID int64, branch string) error {
 }
 
 // RemoveOldDeletedBranches removes old deleted branches
-func RemoveOldDeletedBranches(ctx context.Context, olderThan time.Duration) {
+func RemoveOldDeletedBranches(olderThan time.Duration) {
```
Comment (zeripath) [inline models/branches.go:583]: Please don't remove the context here. 

We need to keep this because we're going to start pushing context through the stack in order to simplify things like process hierarchies, logging and db transactions.

File: models/repo_generate.go [hunk 7]
```diff
@@ -80,7 +80,7 @@ func GenerateTopics(ctx context.Context, templateRepo, generateRepo *Repository)
 }
 
 // GenerateGitHooks generates git hooks from a template repository
-func GenerateGitHooks(ctx context.Context, templateRepo, generateRepo *Repository) error {
+func GenerateGitHooks(templateRepo, generateRepo *Repository) error {
```
Comment (zeripath) [inline models/repo_generate.go:83]: Please don't remove context here.

File: modules/git/commit_info_nogogit.go [hunk 8]
```diff
@@ -99,7 +99,7 @@ func (tes Entries) GetCommitsInfo(ctx context.Context, commit *Commit, treePath
 	return commitsInfo, treeCommit, nil
 }
 
-func getLastCommitForPathsByCache(ctx context.Context, commitID, treePath string, paths []string, cache *LastCommitCache) (map[string]*Commit, []string, error) {
+func getLastCommitForPathsByCache(commitID, treePath string, paths []string, cache *LastCommitCache) (map[string]*Commit, []string, error) {
```
Comment (zeripath) [inline modules/git/commit_info_nogogit.go:102]: Please don't remove context

File: modules/git/commit_info_nogogit.go [hunk 9]
```diff
@@ -31,7 +31,7 @@ func (tes Entries) GetCommitsInfo(ctx context.Context, commit *Commit, treePath
 	var revs map[string]*Commit
 	if cache != nil {
 		var unHitPaths []string
-		revs, unHitPaths, err = getLastCommitForPathsByCache(ctx, commit.ID.String(), treePath, entryPaths, cache)
+		revs, unHitPaths, err = getLastCommitForPathsByCache(commit.ID.String(), treePath, entryPaths, cache)
```
Comment (zeripath) [inline modules/git/commit_info_nogogit.go:34]: Please don't remove context

File: modules/markup/common/footnote.go (hunk 2 above)
Comment (wxiaoguang) [reply modules/markup/common/footnote.go:235]: It's still better to keep the names.

`Open(parent ast.Node, ...`
`Continue(node ast.Node, ...`

They have different meanings.

File: modules/git/diff.go (hunk 1 above)
Comment (zeripath) [reply modules/git/diff.go:296]: In this case it's fine. This internal function doesn't need to call cancel or to pay attention to the provided context as if the context is cancelled the stdout pipe should get closed and then the whole thing will stop.

File: modules/lfs/filesystem_client.go [hunk 10]
```diff
@@ -39,7 +39,7 @@ func (c *FilesystemClient) objectPath(oid string) string {
 }
 
 // Download reads the specific LFS object from the target path
-func (c *FilesystemClient) Download(ctx context.Context, objects []Pointer, callback DownloadCallback) error {
+func (c *FilesystemClient) Download(_ context.Context, objects []Pointer, callback DownloadCallback) error {
```
Comment (zeripath) [inline modules/lfs/filesystem_client.go:42]: The lack of use of ctx here is possibly an error and it likely should be being checked before callback is called.

File: modules/lfs/filesystem_client.go [hunk 11]
```diff
@@ -58,7 +58,7 @@ func (c *FilesystemClient) Download(ctx context.Context, objects []Pointer, call
 }
 
 // Upload writes the specific LFS object to the target path
-func (c *FilesystemClient) Upload(ctx context.Context, objects []Pointer, callback UploadCallback) error {
+func (c *FilesystemClient) Upload(_ context.Context, objects []Pointer, callback UploadCallback) error {
```
Comment (zeripath) [inline modules/lfs/filesystem_client.go:61]: Similarly here.

File: modules/markup/common/footnote.go (hunk 2 above)
Comment (chavacava) [reply modules/markup/common/footnote.go:235]: > They have different meanings.

These parameters mean nothing to those methods (because they are unused)
What is the interest of documenting (putting a name to) something that is not used?

As user of the method it seems, to me, more clear a `_` than an actual name. 
`_` means "_do not take care of this parameter, it has no influence in the method behavior_", then as user of the method I know (instantly) that I can pass anything as argument. Contrary, if the parameter has a name, I will need to take care of providing an argument that fits the expectations (e.g. a valid `context` or an actual _parent_ `ast.Node`)... for nothing because the parameter is not used.
